content/languages/cpp/authoring/index.md

```diff
@@ -0,0 +1,378 @@
+
+## General info
+
+Any technical information related to the C++ setup on Codewars can be found on the [C++ reference](/languages/cpp/) page (language versions, available libraries, and setup of the code runner).
+
+
+## Description
+
+C++-specific code can be inserted with the `cpp` flag in [sequential code blocks](/references/markdown/extensions/#sequential-code-blocks):
+
+~~~
+```cpp
+
+...your code here...
+
+```
+~~~
+
+C++-specific paragraphs can be inserted with [language conditional rendering](/references/markdown/extensions/#conditional-rendering):
+
+```
+~~~if:cpp
+
+...text visible only for C++ description...
+
+~~~
+
+~~~if-not:cpp
+
+...text not visible in C++ description...
+
+~~~
+```
+
+## Tasks and Requirements
+
+Some concepts don't always translate well to or from C++. C++ allows for a variety of paradigms and techniques, many of which are not necessarily suited to other languages. For this reason, certain constructs should be avoided and some kata just should not be translated to C++.
+- C++ is statically typed, so any task that depends on dynamic typing can be difficult to translate into C++, and attempts to force a C++ kata to reflect a dynamically typed interface can lead to ideas that enforce a really bad design.
+- There are only a few additional libraries available to the C++ runner, so kata that take advantage of specialized packages installed for other languages become much more difficult to implement in C++.
+- Many features of C++ do not have direct equivalents in other popular languages: mixture of allowed paradigms, template meta-programming, unmanaged memory backed by RAII, native access to the platform and runtime, and many others. C++ kata which rely on them can be difficult to translate to other languages.
+
+
+## Kata snippets and template
+
+Due to the complicated compilation model of C++ code, paired with the fact that Codewars strictly enforces the names, amount, and meaning of source files, the code runner uses a `[template]()` (TODO: document C++ runner template) to concatenate all kata snippets into a single translation unit. This fact has various consequences on the code, some positive, and some negative: namespace pollution, symbols introduced by one snippet being visible in other snippets, etc. However, in the majority of cases, it does not affect kata in any way, and whenever possible, this behavior should be treated as an implementation detail of the C++ code runner. Kata snippets should be treated as separate source files, if possible.
+
+One consequence of using a template is that the signature of a solution function can be modified by the user in a way which can affect tests. To avoid the possibility of users tampering with the prototype of a solution function, it's recommended to re-declare it in the tests snippet.
```

Comment:
Redeclaration is questionable. It can be convenient to choose between accepting input by reference and by value (when a solution needs a copy anyway) and I think the only thing it prevents is lvalue-rvalue overloading that can be used to distinguish between fixed and random tests, which isn't really useful except when the tests are bad in the first place.

Reply:
You are not a first person to raise this issue, but as much as I can see where you guys are coming from, I do not like it ;)

My main idea behind this point was to prevent situations when a user changes const ref to non-const ref and mutates inputs to modify behavior of test cases. Or when user changes a narrow type to a wider type to work around some potential difficulties of solution.

On a side not related to kata, I do not like the idea of changing the interface just for a sake of an implementation detail. I know this technique and I know that it's used, I am just not sure whether a public interface of a solution is a good place to apply it.

I will think how to reword this part. Maybe something along the lines that _"make sure that users won't be able to break tests by modifying a signature of solution function"_, or something like that?

Reply:
```suggestion
One consequence of using a template is that the signature of a solution function can be modified by the user in a way which can affect tests. Make sure that tests are not possible to work around by users tampering with the prototype of the solution function. One possible way to ensure this is to re-declare the solution function in the tests snippet.
```

Do you think it's better now?